S3Request.gs

```diff
@@ -135,128 +157,259 @@ S3Request.prototype.execute = function(options) {
   if (this.getContentType()) {
     params.contentType = this.getContentType();
   }
-  
-  var response = UrlFetchApp.fetch(this.getUrl(), params);
 
+  var response = UrlFetchApp.fetch(this.getUrl(), params);
 
-  
   //debugging stuff
-  var request = UrlFetchApp.getRequest(this.getUrl(), params);  
-
+  var request = UrlFetchApp.getRequest(this.getUrl(), params);
 
   //Log request and response
   this.lastExchangeLog = this.service.logExchange_(request, response);
   if (options.logRequests) {
     Logger.log(this.service.getLastExchangeLog());
   }
-  
+
   //used in case you want to peak at the actual raw HTTP request coming out of Google's UrlFetchApp infrastructure
   if (options.echoRequestToUrl) {
-    UrlFetchApp.fetch(options.echoRequestToUrl, params); 
+    UrlFetchApp.fetch(options.echoRequestToUrl, params);
   }
-  
+
   //check for error codes (AWS uses variants of 200s for flavors of success)
   if (response.getResponseCode() > 299) {
     //convert XML error response from AWS into JS object, and give it a name
     var error = {};
     error.name = "AwsError";
     try {
       var errorXmlElements = XmlService.parse(response.getContentText()).getRootElement().getChildren();
-    
       for (i in errorXmlElements) {
-        var name = errorXmlElements[i].getName(); 
+        var name = errorXmlElements[i].getName();
         name = name.charAt(0).toLowerCase() + name.slice(1);
         error[name] = errorXmlElements[i].getText();
       }
-      error.toString = function() { return "AWS Error - "+this.code+": "+this.message; }; 
-     
+      error.toString = function() { return "AWS Error - "+this.code+": "+this.message; };
       error.httpRequestLog = this.service.getLastExchangeLog();
     } catch (e) {
       //error parsing XML error response from AWS (will obscure actual error)
- 
       error.message = "AWS returned HTTP code " + response.getResponseCode() + ", but error content could not be parsed."
-      
       error.toString = function () { return this.message; };
-      
       error.httpRequestLog = this.service.getLastExchangeLog();
     }
-    
     throw error;
   }
-  
+
   return response;
 };
 
+S3Request.prototype.addAuthorization = function(credentials, date) {
+  var datetime = date.toISOString().replace(/[:\-]|\.\d{3}/g, '');
+  if (this.isPresigned()) {
+    this.updateForPresigned(credentials, datetime);
+  } else {
+    this.addHeaders(credentials, datetime);
+  }
+  this.headers['Authorization'] = this.authorization(credentials, datetime)
+}
+
+S3Request.prototype.addHeaders = function (credentials, datetime) {
+  this.headers['X-Amz-Date'] = datetime;
```

Comment:
probably not relevant, but AWS example shows this in lowercase (`x-amz-date`), and I think it's worth being consistent at least amongst all `x-amz` headers.

Reply:
I can agree the consistent rule. But the master (AWS official) code [v4.js](https://github.com/aws/aws-sdk-js/blob/master/lib/signers/v4.js#L38-L43) uses this camelcase like rule.
We have to select which rule we use.

Reply:
OK, I see. well then I have complaints about the quality of AMZN's code as well, which I guess isn't so surprising. I'm fine to leave it. should reference that you've ported that code to here in a comment somewhere to be clear.

Reply:
OK. I add reference comment in this code.